// @consid-germany/gates/lib/cross-region-string-ref.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ssm = require("aws-cdk-lib/aws-ssm");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const constructs_1 = require("constructs");
const ssm_parameter_reader_1 = require("./ssm-parameter-reader");
const CUSTOM_RESOURCE_TYPE = "Custom::CrossRegionStringRef";
class CrossRegionStringRef extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.stackOfConstruct = aws_cdk_lib_1.Stack.of(props.constructInOtherRegion);
        const parameterName = `/${this.cdkExportsParameterPrefix}/${this.stackOfConstruct.stackName}/${props.constructInOtherRegion.node.path.replace(/[^/\w.-]/g, "-")}/${id}`;
        new ssm.StringParameter(props.constructInOtherRegion, id, {
            parameterName,
            stringValue: props.value(props.constructInOtherRegion),
        });
        this.ssmParameterReader = new ssm_parameter_reader_1.SSMParameterReader(this, id, {
            parameterName,
            region: this.stackOfConstruct.region,
            resourceType: CUSTOM_RESOURCE_TYPE,
        });
    }
    get value() {
        return this.ssmParameterReader.value;
    }
    get cdkExportsParameterPrefix() {
        return `cdk/${this.stackOfConstruct.synthesizer.bootstrapQualifier}/exports`;
    }
}
exports.default = CrossRegionStringRef;
