// @consid-germany/gates/lib/gates.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Gates = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constructs_1 = require("constructs");
const cdk = require("aws-cdk-lib");
const wafv2 = require("aws-cdk-lib/aws-wafv2");
const route53 = require("aws-cdk-lib/aws-route53");
const route53_targets = require("aws-cdk-lib/aws-route53-targets");
const acm = require("aws-cdk-lib/aws-certificatemanager");
const lambda = require("aws-cdk-lib/aws-lambda");
const logs = require("aws-cdk-lib/aws-logs");
const apigatewayv2 = require("aws-cdk-lib/aws-apigatewayv2");
const apigatewayv2_integrations = require("aws-cdk-lib/aws-apigatewayv2-integrations");
const dynamodb = require("aws-cdk-lib/aws-dynamodb");
const secretsmanager = require("aws-cdk-lib/aws-secretsmanager");
const cloudfront = require("aws-cdk-lib/aws-cloudfront");
const s3 = require("aws-cdk-lib/aws-s3");
const s3_deployment = require("aws-cdk-lib/aws-s3-deployment");
const global_stack_1 = require("./global-stack");
const cross_region_string_ref_1 = require("./cross-region-string-ref");
const path = require("path");
const apigatewayv2_authorizers = require("aws-cdk-lib/aws-apigatewayv2-authorizers");
const SCOPE_CLOUDFRONT = "CLOUDFRONT";
const DEFAULT_APP_NAME = "gates";
const DEFAULT_GITHUB_SUBDOMAIN = "github";
const X_VERIFY_ORIGIN_HEADER_NAME = "x-verify-origin";
class Gates extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const { appName = DEFAULT_APP_NAME } = props;
        this.stack = cdk.Stack.of(this);
        this.globalStack = global_stack_1.default.getOrCreate(this, {
            stackName: props.globalStackName || `${this.stack.stackName}-global`,
            tags: this.stack.tags.tagValues(),
        });
        const hostedZone = this.lookupHostedZone(props.domain);
        const gatesTable = this.createGatesTable(appName);
        const apiFunction = this.createApiFunction(appName, gatesTable, props.demoMode);
        const verifyOriginSecret = this.createVerifyOriginSecret(appName);
        const verifyOriginAuthFunction = this.createVerifyOriginAuthFunction(appName, verifyOriginSecret);
        const gatesApi = this.createGatesApi(appName, apiFunction, verifyOriginAuthFunction);
        if (props.gitHubApi !== undefined) {
            const gitHubJwtAuthFunction = this.createGitHubJwtAuthFunction(appName, props.gitHubApi.allowedSubPatterns);
            this.createGitHubApi(appName, apiFunction, gitHubJwtAuthFunction, hostedZone, props.domain);
        }
        const frontendAssetsBucket = this.createFrontendAssetsBucket(props.frontendAssetsBucketName);
        const webDistribution = this.createWebDistribution(appName, frontendAssetsBucket, gatesApi, verifyOriginSecret, hostedZone, props.domain, props.ipAllowList);
        this.createVerifyOriginSecretRotation(verifyOriginSecret, webDistribution, gatesApi);
        this.createFrontendAssetsDeployment(frontendAssetsBucket, webDistribution);
    }
    lookupHostedZone(domain) {
        if (domain === undefined) {
            return undefined;
        }
        return route53.HostedZone.fromLookup(this, "HostedZone", {
            domainName: domain.zoneDomainName || domain.domainName,
        });
    }
    createGitHubJwtAuthFunction(appName, allowedSubPatterns) {
        return new lambda.Function(this, "GitHubJwtAuthFunction", {
            functionName: `${appName}-github-jwt-auth`,
            runtime: lambda.Runtime.NODEJS_20_X,
            code: lambda.Code.fromAsset(path.join(__dirname, "..", "build", "function", "github-jwt-authorizer")),
            handler: "index.handler",
            logRetention: logs.RetentionDays.ONE_WEEK,
            environment: {
                ALLOWED_SUB_PATTERNS: JSON.stringify(allowedSubPatterns),
            },
        });
    }
    createGitHubApi(appName, apiFunction, gitHubJwtAuthFunction, hostedZone, domain) {
        const apiFunctionIntegration = new apigatewayv2_integrations.HttpLambdaIntegration("ApiFunctionGitHubHttpApiIntegration", apiFunction);
        const gitHubJwtAuthorizer = new apigatewayv2_authorizers.HttpLambdaAuthorizer("GitHubJwtAuthorizer", gitHubJwtAuthFunction, {
            responseTypes: [apigatewayv2_authorizers.HttpLambdaResponseType.SIMPLE],
        });
        const httpApi = new apigatewayv2.HttpApi(this, "GitHubHttpApi", {
            apiName: `${appName}-github-api`,
            defaultDomainMapping: this.createGitHubDomainMapping(hostedZone, domain),
        });
        httpApi.addRoutes({
            integration: apiFunctionIntegration,
            authorizer: gitHubJwtAuthorizer,
            path: "/api/gates/{group}/{service}/{environment}/state",
            methods: [apigatewayv2.HttpMethod.GET]
        });
    }
    createGitHubDomainMapping(hostedZone, domain) {
        if (hostedZone === undefined || domain === undefined) {
            return undefined;
        }
        const gitHubApiDomainName = `${domain.gitHubApiSubdomain || DEFAULT_GITHUB_SUBDOMAIN}.${domain.domainName}`;
        const gitHubSubdomainCertificate = new acm.Certificate(this, "GitHubHttpApiCertificate", {
            domainName: gitHubApiDomainName,
            validation: acm.CertificateValidation.fromDns(hostedZone),
        });
        const gitHubHttpApiDomainName = new apigatewayv2.DomainName(this, "GitHubHttpApiDomain", {
            domainName: gitHubApiDomainName,
            certificate: gitHubSubdomainCertificate,
        });
        new route53.ARecord(this, "GitHubHttpApiARecord", {
            recordName: gitHubApiDomainName,
            target: route53.RecordTarget.fromAlias(new route53_targets.ApiGatewayv2DomainProperties(gitHubHttpApiDomainName.regionalDomainName, gitHubHttpApiDomainName.regionalHostedZoneId)),
            zone: hostedZone,
        });
        return {
            domainName: gitHubHttpApiDomainName
        };
    }
    createFrontendAssetsDeployment(frontendAssetsBucket, webDistribution) {
        new s3_deployment.BucketDeployment(this, "BucketDeployment", {
            sources: [s3_deployment.Source.asset(path.join(__dirname, "..", "build", "ui"))],
            destinationBucket: frontendAssetsBucket,
            distribution: webDistribution,
        });
    }
    createFrontendAssetsBucket(bucketName) {
        return new s3.Bucket(this, "FrontendAssetsBucket", {
            bucketName,
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_ENFORCED,
            autoDeleteObjects: true,
            removalPolicy: cdk.RemovalPolicy.DESTROY,
        });
    }
    createWebDistribution(appName, frontendAssetsBucket, httpApi, verifyOriginSecret, hostedZone, domain, ipAllowList) {
        const cloudfrontOAI = new cloudfront.OriginAccessIdentity(this, "OriginAccessIdentity");
        frontendAssetsBucket.grantRead(cloudfrontOAI);
        const webDistribution = new cloudfront.CloudFrontWebDistribution(this, "WebDistribution", {
            webACLId: this.createGlobalWebAcl(appName, ipAllowList),
            enableIpV6: false,
            viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
            viewerCertificate: this.createViewerCertificate(domain),
            originConfigs: [
                {
                    customOriginSource: {
                        domainName: `${httpApi.apiId}.execute-api.${this.stack.region}.amazonaws.com`,
                        originHeaders: {
                            [X_VERIFY_ORIGIN_HEADER_NAME]: verifyOriginSecret.secretValue.unsafeUnwrap(),
                        },
                    },
                    behaviors: [
                        {
                            pathPattern: "/api",
                            allowedMethods: cloudfront.CloudFrontAllowedMethods.ALL,
                            defaultTtl: cdk.Duration.seconds(0),
                        },
                        {
                            pathPattern: "/api/*",
                            allowedMethods: cloudfront.CloudFrontAllowedMethods.ALL,
                            defaultTtl: cdk.Duration.seconds(0),
                        },
                    ],
                },
                {
                    s3OriginSource: {
                        s3BucketSource: frontendAssetsBucket,
                        originAccessIdentity: cloudfrontOAI,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            allowedMethods: cloudfront.CloudFrontAllowedMethods.GET_HEAD_OPTIONS,
                        },
                    ],
                },
            ],
        });
        if (domain !== undefined && hostedZone !== undefined) {
            new route53.ARecord(this, "ARecord", {
                recordName: domain.domainName,
                target: route53.RecordTarget.fromAlias(new route53_targets.CloudFrontTarget(webDistribution)),
                zone: hostedZone,
            });
        }
        return webDistribution;
    }
    createGatesApi(appName, apiFunction, verifyOriginAuthFunction) {
        const apiFunctionIntegration = new apigatewayv2_integrations.HttpLambdaIntegration("ApiFunctionApiIntegration", apiFunction);
        return new apigatewayv2.HttpApi(this, "HttpApi", {
            apiName: `${appName}-api`,
            defaultIntegration: apiFunctionIntegration,
            defaultAuthorizer: new apigatewayv2_authorizers.HttpLambdaAuthorizer("VerifyOriginAuthorizer", verifyOriginAuthFunction, {
                responseTypes: [apigatewayv2_authorizers.HttpLambdaResponseType.SIMPLE],
                identitySource: [`$request.header.${X_VERIFY_ORIGIN_HEADER_NAME}`],
            }),
        });
    }
    createVerifyOriginSecretRotation(verifyOriginSecret, webDistribution, httpApi) {
        const verifyOriginSecretRotationFunction = new lambda.Function(this, "VerifyOriginSecretRotationFunction", {
            functionName: `${verifyOriginSecret.secretName}-rotation`,
            runtime: lambda.Runtime.NODEJS_20_X,
            code: lambda.Code.fromAsset(path.join(__dirname, "..", "build", "function", "verify-origin-secret-rotation")),
            handler: "index.handler",
            logRetention: logs.RetentionDays.ONE_WEEK,
            timeout: cdk.Duration.seconds(30),
            environment: {
                CLOUDFRONT_DISTRIBUTION_ID: webDistribution.distributionId,
                X_VERIFY_ORIGIN_HEADER_NAME,
                ORIGIN_TEST_URL: `https://${httpApi.apiId}.execute-api.${this.stack.region}.amazonaws.com/api/`,
            },
        });
        webDistribution.grant(verifyOriginSecretRotationFunction, "cloudfront:GetDistribution", "cloudfront:GetDistributionConfig", "cloudfront:UpdateDistribution");
        verifyOriginSecret.addRotationSchedule("RotationSchedule", {
            rotationLambda: verifyOriginSecretRotationFunction,
            automaticallyAfter: cdk.Duration.days(1),
        });
    }
    createVerifyOriginAuthFunction(appName, verifyOriginSecret) {
        const verifyOriginAuthFunction = new lambda.Function(this, "VerifyOriginAuthFunction", {
            functionName: `${appName}-verify-origin-auth`,
            runtime: lambda.Runtime.NODEJS_20_X,
            code: lambda.Code.fromAsset(path.join(__dirname, "..", "build", "function", "verify-origin-authorizer")),
            handler: "index.handler",
            logRetention: logs.RetentionDays.ONE_WEEK,
            environment: {
                SECRET_ID: verifyOriginSecret.secretName,
                X_VERIFY_ORIGIN_HEADER_NAME,
            },
        });
        verifyOriginSecret.grantRead(verifyOriginAuthFunction);
        return verifyOriginAuthFunction;
    }
    createVerifyOriginSecret(appName) {
        return new secretsmanager.Secret(this, "VerifyOriginSecret", {
            secretName: `${appName}-verify-origin-secret`,
            generateSecretString: {
                excludePunctuation: true,
            },
        });
    }
    createApiFunction(appName, gatesTable, demoMode) {
        const apiFunction = new lambda.Function(this, "ApiFunction", {
            functionName: `${appName}-api`,
            runtime: lambda.Runtime.PROVIDED_AL2023,
            architecture: lambda.Architecture.ARM_64,
            code: lambda.Code.fromAsset(path.join(__dirname, "..", "build", "api")),
            handler: "provided",
            environment: {
                GATES_DYNAMO_DB_TABLE_NAME: gatesTable.tableName,
                ...(demoMode && { DEMO_MODE: "true" }),
            },
            logRetention: logs.RetentionDays.ONE_WEEK,
        });
        gatesTable.grantReadWriteData(apiFunction);
        return apiFunction;
    }
    createViewerCertificate(domain) {
        if (domain === undefined) {
            return undefined;
        }
        return cloudfront.ViewerCertificate.fromAcmCertificate(this.createGlobalCertificate(domain), {
            aliases: [domain.domainName],
        });
    }
    createGatesTable(appName) {
        return new dynamodb.TableV2(this, "GatesTable", {
            tableName: `${appName}`,
            partitionKey: { name: "group", type: dynamodb.AttributeType.STRING },
            sortKey: { name: "service_environment", type: dynamodb.AttributeType.STRING },
        });
    }
    createGlobalCertificate(domain) {
        const hostedZone = route53.HostedZone.fromLookup(this.globalStack, "HostedZone", {
            domainName: domain.zoneDomainName || domain.domainName,
        });
        const certificate = new acm.Certificate(this.globalStack, "GlobalCertificate", {
            domainName: domain.domainName,
            validation: acm.CertificateValidation.fromDns(hostedZone),
        });
        const certificateArn = new cross_region_string_ref_1.default(this, "GlobalCertificateArn", {
            constructInOtherRegion: certificate,
            value: (certificate) => certificate.certificateArn,
        }).value;
        return acm.Certificate.fromCertificateArn(this, "GlobalCertificate", certificateArn);
    }
    createGlobalWebAcl(appName, ipAllowList) {
        if (ipAllowList === undefined) {
            return undefined;
        }
        const ipSet = new wafv2.CfnIPSet(this.globalStack, "IpSet", {
            name: `${appName}-ip-allow-list`,
            addresses: [...ipAllowList],
            ipAddressVersion: "IPV4",
            scope: "CLOUDFRONT",
        });
        const ipAllowListRule = {
            name: `${appName}-waf-ip-allow-list-rule`,
            visibilityConfig: {
                cloudWatchMetricsEnabled: true,
                metricName: `${appName}-waf-ip-allow-list`,
                sampledRequestsEnabled: true,
            },
            priority: 0,
            action: { allow: {} },
            statement: {
                ipSetReferenceStatement: {
                    arn: ipSet.attrArn,
                },
            },
        };
        const webAcl = new wafv2.CfnWebACL(this.globalStack, "WebAcl", {
            name: `${appName}-waf`,
            defaultAction: { block: {} },
            visibilityConfig: {
                cloudWatchMetricsEnabled: true,
                metricName: `${appName}-waf`,
                sampledRequestsEnabled: true,
            },
            scope: SCOPE_CLOUDFRONT,
            rules: [ipAllowListRule],
        });
        return new cross_region_string_ref_1.default(this, "WebAclArn", {
            constructInOtherRegion: webAcl,
            value: (webAcl) => webAcl.attrArn,
        }).value;
    }
}
exports.Gates = Gates;
_a = JSII_RTTI_SYMBOL_1;
Gates[_a] = { fqn: "@consid-germany/gates.Gates", version: "1.0.0" };
