// @consid-germany/gates/lib/ssm-parameter-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSMParameterReader = void 0;
const aws_cdk_lib_1 = require("aws-cdk-lib");
const CustomResource = require("aws-cdk-lib/custom-resources");
class SSMParameterReader extends CustomResource.AwsCustomResource {
    constructor(scope, id, props) {
        const onUpdate = {
            service: "SSM",
            action: "getParameter",
            parameters: {
                Name: props.parameterName,
            },
            region: props.region,
            physicalResourceId: CustomResource.PhysicalResourceId.of(Date.now().toString()),
        };
        const policy = CustomResource.AwsCustomResourcePolicy.fromSdkCalls({
            resources: [
                aws_cdk_lib_1.Arn.format({
                    service: "ssm",
                    region: props.region,
                    resource: "parameter",
                    resourceName: removeLeadingSlash(props.parameterName),
                }, aws_cdk_lib_1.Stack.of(scope)),
            ],
        });
        super(scope, id, { onUpdate, policy, resourceType: props.resourceType });
    }
    get value() {
        return this.getResponseField("Parameter.Value");
    }
}
exports.SSMParameterReader = SSMParameterReader;
function removeLeadingSlash(value) {
    return value.slice(0, 1) == "/" ? value.slice(1) : value;
}
